src/librustc_mir/dataflow/impls/borrows.rs

```diff
@@ -273,6 +273,246 @@ impl<'a, 'gcx, 'tcx> Borrows<'a, 'gcx, 'tcx> {
+            /// compiler since two-phase borrows are only currently supported for compiler-
+            /// generated code. More precisely, we only allow two-phase borrows for:
+            ///   - Function calls (fn some_func(&mut self, ....))
+            ///   - *Assign operators (a += b -> fn add_assign(&mut self, other: Self))
+            /// See
+            ///   - https://github.com/rust-lang/rust/issues/48431
+            /// for detailed design notes.
+            /// See the TODO in the body of the function for notes on extending support to more
+            /// general two-phased borrows.
+            fn compute_activation_location(&self,
+                                           start_location: Location,
+                                           assigned_place: &mir::Place<'tcx>,
+                                           region: Region<'tcx>,
+                                           kind: mir::BorrowKind) -> Location {
+                debug!("Borrows::compute_activation_location({:?}, {:?}, {:?})",
+                       start_location,
+                       assigned_place,
+                       region);
+                if !self.allow_two_phase_borrow(kind) {
+                    debug!("  -> {:?}", start_location);
+                    return start_location;
+                }
+
+                // Perform the DFS.
+                // `stack` is the stack of locations still under consideration
+                // `found_use` is an Option that becomes Some when we find a use
+                let mut stack = vec![start_location];
+                let mut found_use = None;
+                while let Some(curr_loc) = stack.pop() {
+                    let block_data = &self.mir.basic_blocks()
+                        .get(curr_loc.block)
+                        .unwrap_or_else(|| {
+                            panic!("could not find block at location {:?}", curr_loc);
+                        });
+
+                    if self.region_terminated_after(region, curr_loc) {
+                        // No need to process this statement.
+                        // It's either an EndRegion (and thus couldn't use assigned_place) or not
+                        // contained in the NLL region and thus a use would be invalid
+                        continue;
+                    }
+
+                    if self.location_contains_use(curr_loc, assigned_place) {
+                        // TODO: Handle this case a little more gracefully. Perhaps collect
+                        // all uses in a vector, and find the point in the CFG that dominates
+                        // all of them?
+                        // Right now this is sufficient though since there should only be exactly
+                        // one borrow-activating use of the borrow.
+                        assert!(found_use.is_none(), "Found secondary use of place");
+                        found_use = Some(curr_loc);
```

Comment:
Hmm, I would have expected you to `continue` to the top of the `while let` here, rather than pushing the points in the control-flow that suceed the use that you have found. I don't mind pushing those points (traversing them represents a small bit of extra effort, I suspect), I just found it a little surprising.

Reply:
I agree, that should be sufficient. I wrote it this way in case we wanted to support general two-phase but if that's not of interest (as indicated by some comments above) I'll use the continue instead.